### PR TITLE
[AO][bugfix] Fixing FX QAT but for untraceable modules

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74277

Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

### CODE REVIEW

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

File: torch/ao/quantization/quantize_fx.py [hunk 1]
```diff
@@ -134,13 +139,22 @@ def __init__(
...
     def is_leaf_module(self, m: torch.nn.Module, module_qualified_name: str) -> bool:
+        # children of untraceable modules need to be identified so that
+        # qconfigs are not propagated to them incorrectly later.
+        if (
+            module_qualified_name in self.non_traceable_module_names
+            or type(m) in self.non_traceable_module_classes
+        ):
+            for name, child in m.named_children():
+                if not hasattr(child, "qconfig"):
+                    child.qconfig = None  # type: ignore[assignment]
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize_fx.py:150]: I feel ideally we should set qconfig in one place, to keep code easier to maintain and reason about, can we do this in prepare?

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

File: torch/ao/quantization/quantize.py [hunk 2]
```diff
@@ -61,10 +64,15 @@ def _propagate_qconfig_helper(module, qconfig_dict,
 
     for name, child in module.named_children():
         module_prefix = prefix + '.' + name if prefix else name
-        _propagate_qconfig_helper(child, qconfig_dict,
-                                  qconfig_with_device_check, module_prefix)
+        if prepare_custom_config_dict is None or not (
```
Comment (vkuzo) [inline torch/ao/quantization/quantize.py:67]: can we add a comment with a human readable description on what this is doing?

File: torch/ao/quantization/quantize.py [hunk 3]
```diff
@@ -61,10 +64,15 @@ def _propagate_qconfig_helper(module, qconfig_dict,
...
+            name in prepare_custom_config_dict.get("non_traceable_module_name", [])
+            or type(child) in prepare_custom_config_dict.get("non_traceable_module_class", [])
+        ):
+            _propagate_qconfig_helper(
+                child, qconfig_dict, qconfig_with_device_check, module_prefix
+            )
 
-def propagate_qconfig_(module, qconfig_dict=None):
+def propagate_qconfig_(module, qconfig_dict=None, prepare_custom_config_dict=None):
```
Comment (vkuzo) [inline torch/ao/quantization/quantize.py:75]: could we add the arg to the docstring too?

**Review by vkuzo — APPROVED** [review]
looks great!

**Comment by jerryzh168** [comment]
looks great, do we want to add support for custom module as well in a followup PR?

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

**Pull request closed without merging**